### PR TITLE
[Doc] [Core] [runtime env] Add runtime env doc

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds documentation for runtime envs.  Some code samples from these docs are drawn from a script which is tested in CI.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
Is it possible to have all of the code samples tested in CI?

File: doc/source/advanced.rst [hunk 1]
```diff
@@ -423,33 +423,75 @@ To get information about the current available resource capacity of your cluster
...
+
+  - Example: ``"./environment.yml"``
+
+  - Example: ``"pytorch_p36"``
+
+- ``env_vars`` (Dict[str, str]): Environment variables to set.
+
+  - Example: ``{"OMP_NUM_THREADS": "32", "TF_WARNINGS": "none"}``
+
+The runtime env is inheritable, so it will apply to all tasks/actors within a job and all child tasks/actors of a task or actor, once set.
```
Comment (edoakes) [inline doc/source/advanced.rst:494]: I feel like all of this should just be coming from the docstring of runtime_env -- is there a way we can just do the documentation for the individual options there and embed it here?

File: doc/source/advanced.rst [hunk 2]
```diff
@@ -423,33 +423,75 @@ To get information about the current available resource capacity of your cluster
...
-must all use the same Python minor version (e.g., Python 3.8).  Note that this feature is currently unavailable on Windows.
+The ``runtime_env`` option is a (JSON-serializable) dictionary that can be passed as an option to tasks and actors, and to ``ray.init()`` and ``ray.client().connect()``. The runtime environment defines the dependencies required for your workload.
 
-To start a specific task or an actor in an existing conda environment, pass in the environment name to your task or 
-actor via the ``runtime_env`` parameter as follows:
+Examples:
 
 .. code-block:: python
+    
+    runtime_env = {"working_dir": "local_project.zip", "pip": ["chess==1.5.0"]}
```
Comment (iycheng) [inline doc/source/advanced.rst:438]: Rright now we don't support working_dir as a zip.
It should be a local dir.

File: doc/source/advanced.rst (hunk 1 above)
Comment (architkulkarni) [reply doc/source/advanced.rst:494]: I agree this makes sense as a long term solution, but right now runtime_env is a simple Python dict, so it doesn't have a docstring.  The RuntimeEnvDict class in the code has a docstring, but it's an internal object so I don't think we want to expose it.  I looked for a way for Sphinx to only grab the docstring of RuntimeEnvDict without having the class name appearing in the docs, but couldn't find one.

**Comment by architkulkarni** [comment]
I added a test for the conda runtime_env code sample, both for the job_config case and the per-task and per-actor case.  I verified that the test is running in CI: https://buildkite.com/ray-project/ray-builders-pr/builds/6762#e4425484-2aae-42bf-be9a-a9c63a37a398/6-10401

Adding a test for the Ray Client and `working_dir` example that's compatible with `.. literalinclude::` without causing awkward indentation in the code samples will require some more thought, and I will include this in a followup PR.  I've made an issue here https://github.com/ray-project/ray/issues/16343 to track the desired outcome, which is that all sample code in the runtime_env docs should be run in CI.

**Comment by AmeerHajAli** [comment]
@architkulkarni , when you merge this can you please close "Missing documentation for runtime envs in open source anyscale/product#4680 "

**Review by edoakes — APPROVED** [review]

**Pull request merged**